File: src/algorithms/pathfinding/maze-generation/prims-maze/sources/prims-maze_test.py
import importlib
from collections import deque

prims_maze_mod = importlib.import_module("prims-maze")
prims_maze = prims_maze_mod.prims_maze


def make_all_walls_grid(rows, cols):
    return [[{"type": "wall"} for _ in range(cols)] for _ in range(rows)]


def bfs_reachable(grid, start, end):
    row_count, col_count = len(grid), len(grid[0])
    visited = [[False] * col_count for _ in range(row_count)]
    queue = deque([start])
    visited[start[0]][start[1]] = True
    while queue:
        row, col = queue.popleft()
        if (row, col) == end:
            return True
        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nr, nc = row + dr, col + dc
            if 0 <= nr < row_count and 0 <= nc < col_count and not visited[nr][nc] and grid[nr][nc]["type"] != "wall":
                visited[nr][nc] = True
                queue.append((nr, nc))
    return False


def test_carves_passages():
    grid = make_all_walls_grid(9, 9)
    grid[1][1]["type"] = "start"
    result = prims_maze(grid, (1, 1))
    assert result["passagesCarved"] > 0


def test_creates_connected_maze():
    grid = make_all_walls_grid(9, 9)
    grid[1][1]["type"] = "start"
    grid[7][7]["type"] = "end"
    prims_maze(grid, (1, 1))
    assert bfs_reachable(grid, (1, 1), (7, 7))


def test_does_not_carve_border_cells():
    grid = make_all_walls_grid(9, 9)
    grid[1][1]["type"] = "start"
    prims_maze(grid, (1, 1))
    for col in range(9):
        assert grid[0][col]["type"] == "wall"
        assert grid[8][col]["type"] == "wall"
    for row in range(9):
        assert grid[row][0]["type"] == "wall"
        assert grid[row][8]["type"] == "wall"


def test_start_cell_is_carved():
    grid = make_all_walls_grid(9, 9)
    grid[1][1]["type"] = "start"
    prims_maze(grid, (1, 1))
    assert grid[1][1]["type"] != "wall"


if __name__ == "__main__":
    test_carves_passages()
    test_creates_connected_maze()
    test_does_not_carve_border_cells()
    test_start_cell_is_carved()
    print("All tests passed!")

File: src/algorithms/pathfinding/maze-generation/prims-maze/PrimsMaze_test.java
import java.util.*;

// javac PrimsMaze.java PrimsMaze_test.java && java -ea PrimsMaze_test
public class PrimsMaze_test {

    static int[][] makeAllWallsGrid(int rows, int cols) {
        int[][] grid = new int[rows][cols];
        for (int[] row : grid) Arrays.fill(row, 1);
        return grid;
    }

    static boolean bfsReachable(int[][] grid, int startRow, int startCol, int endRow, int endCol) {
        int rows = grid.length, cols = grid[0].length;
        boolean[][] visited = new boolean[rows][cols];
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{startRow, startCol});
        visited[startRow][startCol] = true;
        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            if (curr[0] == endRow && curr[1] == endCol) return true;
            for (int[] dir : dirs) {
                int nextRow = curr[0] + dir[0], nextCol = curr[1] + dir[1];
                if (nextRow >= 0 && nextRow < rows && nextCol >= 0 && nextCol < cols
                        && !visited[nextRow][nextCol] && grid[nextRow][nextCol] == 0) {
                    visited[nextRow][nextCol] = true;
                    queue.add(new int[]{nextRow, nextCol});
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        // Test: carves passages
        {
            int[][] grid = makeAllWallsGrid(9, 9);
            grid[1][1] = 0;
            int passagesCarved = PrimsMaze.primsMaze(grid, new int[]{1, 1});
            assert passagesCarved > 0 : "Expected passages carved > 0";
        }

        // Test: creates connected maze
        {
            int[][] grid = makeAllWallsGrid(9, 9);
            grid[1][1] = 0;
            grid[7][7] = 0;
            PrimsMaze.primsMaze(grid, new int[]{1, 1});
            assert bfsReachable(grid, 1, 1, 7, 7) : "Start should reach end";
        }

        // Test: does not carve border cells
        {
            int[][] grid = makeAllWallsGrid(9, 9);
            grid[1][1] = 0;
            PrimsMaze.primsMaze(grid, new int[]{1, 1});
            for (int col = 0; col < 9; col++) {
                assert grid[0][col] == 1 : "Row 0 should remain wall";
                assert grid[8][col] == 1 : "Row 8 should remain wall";
            }
        }

        // Test: start cell is carved
        {
            int[][] grid = makeAllWallsGrid(9, 9);
            grid[1][1] = 0;
            PrimsMaze.primsMaze(grid, new int[]{1, 1});
            assert grid[1][1] == 0 : "Start cell should be carved";
        }

        System.out.println("All tests passed!");
    }
}
